### ISSUE TITLE
Automatically stop LSP server if no buffers are attached

### ISSUE THREAD

**Issue by famiu** [neovim/neovim#14430]
It seems that currently, even if no buffers are attached to an LSP server, the LSP server keeps running which to me seems like a waste of system resources. So it'd be nice if a method was implemented so that LSP servers are closed when no buffers are attached to them.

**Comment by teto** [comment]
I would say it's more the job of the LSP server to throttle itself. If the LSP server finds itself without buffer attached, only to restart it a short while later, it could consume more. Maybe we could start a timeout after we detect no buffer attach and kill only after that timeout expiration.

**Comment by famiu** [comment]
> I would say it's more the job of the LSP server to throttle itself. If the LSP server finds itself without buffer attached, only to restart it a short while later, it could consume more. Maybe we could start a timeout after we detect no buffer attach and kill only after that timeout expiration.

Yes a timeout would be nice. Maybe allow the user to configure it moreover?

**Comment by teto** [comment]
It's low priority. Note that with nvim-lspconfig you can call `:LspStop <server>`.

**Comment by famiu** [comment]
> It's low priority. Note that with nvim-lspconfig you can call `:LspStop <server>`.

**Comment by famiu** [comment]
> It's low priority. Note that with nvim-lspconfig you can call `:LspStop <server>`.

I see. Also sorry for accidentally closing it. I didn't even notice, lol.
But I don't think it'd be too hard to implement, right?

**Comment by mjlbach** [comment]
I'm not sure, what is the advantage of keeping servers open if a user intentionally deletes all buffers the server is attached to?

**Comment by teto** [comment]
Restarting the server can take a while on big projects so if the 0 buffer is just a transitory phase (maybe loading a session closes buffers and open new ones, dunno  ?), we shouldn't kill the server straightaway, hence the timeout.

**Comment by weihanglo** [comment]
I second @teto. Especially in a multi-language project such as Rust binding for a C library. It will be painful to restart rust-analyzer again and again.

**Comment by famiu** [comment]
I agree with @teto, but I still think a configurable timeout is necessary since not closing the server at all can also be a waste of resources

**Comment by mfussenegger** [comment]
Given that neovim core itself doesn't even define any heuristics to start a language server (and there are afaik no plans to change that), adding heuristics to stop a language server also seems out-of-scope. 

I second the other concerns raised, that just because you close the last buffer of a certain file type doesn't mean you're done with the work, and starting language servers has a significant cost due to the indexing they often require.

What we could do is make it easier to stop a server, e.g. by defining a command like:

```lua
  api.nvim_create_user_command(
    'LspStop',
    function(kwargs)
      local name = kwargs.fargs[1]
      for _, client in ipairs(get_clients({ name = name })) do
        client.stop()
      end
    end,
    {
      nargs = 1,
      complete = function()
        return vim.tbl_map(function(c) return c.name end, get_clients())
      end
    }
  )
```

**Comment by famiu** [comment]
> Given that neovim core itself doesn't even define any heuristics to start a language server (and there are afaik no plans to change that), adding heuristics to stop a language server also seems out-of-scope.
> 
> I second the other concerns raised, that just because you close the last buffer of a certain file type doesn't mean you're done with the work, and starting language servers has a significant cost due to the indexing they often require.
> 
> What we could do is make it easier to stop a server, e.g. by defining a command like:
> 
> ```lua
>   api.nvim_create_user_command(
>     'LspStop',
>     function(kwargs)
>       local name = kwargs.fargs[1]
>       for _, client in ipairs(get_clients({ name = name })) do
>         client.stop()
>       end
>     end,
>     {
>       nargs = 1,
>       complete = function()
>         return vim.tbl_map(function(c) return c.name end, get_clients())
>       end
>     }
>   )
> ```

Fully on-board with this. What are your thoughts on something like `LspRestart` btw? Is that also out of scope?

**Comment by mfussenegger** [comment]
> What are your thoughts on something like LspRestart btw? Is that also out of scope?

I think that's also in scope. I use currently something like this:

```lua
  api.nvim_create_user_command(
    "LspRestart",
    function(kwargs)
      local name = kwargs.fargs[1]
      for _, client in ipairs(get_clients({ name = name })) do
        local bufs = lsp.get_buffers_by_client_id(client.id)
        client.stop()
        vim.wait(30000, function()
          return lsp.get_client_by_id(client.id) == nil
        end)
        local client_id = lsp.start_client(client.config)
        if client_id then
          for _, buf in ipairs(bufs) do
            lsp.buf_attach_client(buf, client_id)
          end
        end
      end
    end,
    {
      nargs = 1,
      complete = function()
        return vim.tbl_map(function(c) return c.name end, get_clients())
      end
    }
  )
```

**Comment by MarcoBuess** [comment]
Could I just implement an `autocmd` on `BufDelete` that checks active clients for if there are any attached buffers and if not stop the client? I mean this is not as fine as it can be, but if the performance cost is negligible this should be pretty easy to implement I think. I'm not yet familiar enough with `neovim` to tell the performance impact. Maybe you guys can share some idears.

**Comment by dfenerski** [comment]
This is quite terrible to use. In a TS setup after a while you just start to get [`inotify` too many file watchers](https://stackoverflow.com/a/55763478/13163112) errors because the `tsserver`s LSP are never stopped after open.

**Comment by mfussenegger** [comment]
See https://github.com/neovim/neovim/issues/14430#issuecomment-1848355381 there are currently no plans to add something like this.
